Add SwiftUI SliderTaskView

DylanCLi opened the pull request:
Hi, I'm the high school student working with @cbaker6 at the NetReconLab. 

I am adding the test cases now.

Close #320 

It can add any Images, or none at all.

Using a default SwfitUI slider
```swift
CareKit.SliderTaskView(taskID: assessmentTask.id, eventQuery: .init(for: date), storeManager: self.storeManager){ controller in
CareKitUI.SliderTaskView(title: Text(controller.viewModel?.title ?? ""),
                                                             detail: Text(controller.viewModel?.detail ?? ""),
                                                             instructions: Text(controller.viewModel?.instructions ?? ""),
                                                             isComplete: controller.viewModel?.isComplete ?? false,
                                                             action: controller.viewModel?.action ?? {},
                                                             minimumImage: minimumImage, maximumImage: maximumImage,
                                                             initialValue: controller.viewModel?.initialValue ?? 5,
                                                             range: 0...10, step: 1,
                                                             useDefaultSlider: true)
}
```

![sliderDemonstrationDefault](https://user-images.githubusercontent.com/56743995/87477239-ccb08000-c5f5-11ea-8c01-db1d6479668c.gif)

Using a custom SwiftUI slider
```swift
CareKit.SliderTaskView(taskID: assessmentTask.id, eventQuery: .init(for: date), storeManager: self.storeManager){ controller in
CareKitUI.SliderTaskView(title: Text(controller.viewModel?.title ?? ""),
                                                             detail: Text(controller.viewModel?.detail ?? ""),
                                                             instructions: Text(controller.viewModel?.instructions ?? ""),
                                                             isComplete: controller.viewModel?.isComplete ?? false,
                                                             action: controller.viewModel?.action ?? {},
                                                             minimumImage: minimumImage, maximumImage: maximumImage,
                                                             initialValue: controller.viewModel?.initialValue ?? 4,
                                                             range: 0...7, step: 1,
                                                             useDefaultSlider: false)
}
```

![sliderDemonstrationCustom](https://user-images.githubusercontent.com/56743995/87477226-c7ebcc00-c5f5-11ea-89b9-70c0812d5a45.gif)

DylanCLi commented on CareKit/CareKit/iOS/Task/SliderTaskView.swift:
````diff
@@ -0,0 +1,199 @@
+/// The card supports styling using `careKitStyle(_:)`.
+///
+/// ```
+///     +-------------------------------------------------------+
````
Was not sure how to generate a visual representation for the slider card. If you point me towards how to generate it I can do it.

gavirawson-apple commented:
Thanks for this amazing work Dylan! I'll be a able to review this soon and add some comments. In the meantime, here is a tip to help with the slider value binding issues you have been running into.

The slider value can live as state on the synchronized view. To read the value and save it to the store, you can pass it through the action that is performed when the button is tapped.

Here is a bit of a simplified example:

```swift
// This is the static view that lives in CareKitUI
struct SliderTaskView: View {

    @Binding
    var value: Double

    // Note that this action provides access to the slider value.
    let action: (_ value: Double) -> Void

    var body: some View {
        VStack {
            Slider(value: _value)
            Button("Log Value") {
                action(value)
            }
        }
    }
}

// This is the synchronized view that lives in CareKit
struct SynchronizedSliderTaskView: View {

    // The value for the slider should live as state in this view.
    // This acts as the source of truth for the slider value.
    @State
    private var value: Double = 1

    var body: some View {
        SliderTaskView(value: $value, action: createViewModel().action)
    }

    // This function is a bit simplified. Ideally you would ask the controller for the view model.
    func createViewModel() -> SliderTaskViewModel {
        .init { value in
            // Save the slider value to the store here
            print("Slider value: \(value)")
        }
    }
}

// You can use the task controller to create a view model.
struct SliderTaskViewModel {
    var action: (_ value: Double) -> Void
}
```

DylanCLi commented on CareKit/CareKit/iOS/Task/Controllers/OCKSliderTaskController.swift:
```diff
@@ -0,0 +1,63 @@
+    private func makeViewModel(from taskEvents: OCKTaskEvents) -> SliderTaskViewModel? {
+        guard !taskEvents.isEmpty else { return nil }
+
+        if let foundValue = taskEvents.first?.first?.outcome?.values.last?.numberValue?.doubleValue {
```
Hi @gavirawson-apple, 
I’m not sure how to correctly pull the right value from taskEvent. For example, if an event is scheduled multiple times a day, what do I do?
Also there is a bug that could be related where when an app is restarted, the slider will default to show the preset value of the @Published var on line 24: 0. This should be a threading issue because the synchronizer gives the right value, but it only sometimes updates the view. Trying: 
DispatchQueue.main.async { [weak self] in
                self?.value = CGFloat(foundValue)
}
on line 41 does not solve the issue but perhaps something similar could.

DylanCLi commented on CareKit/CareKit/iOS/Task/Controllers/OCKSliderTaskController.swift:
```diff
@@ -0,0 +1,63 @@
+    
+    /// Data used to create a `CareKitUI.SliderTaskTaskView`.
+    @Published public private(set) var value: CGFloat = 0 {
+        willSet { objectWillChange.send() }
```
Maybe the DispatchQueue line in the comment below could go in the didSet of this but I was not sure how to implement this

erik-apple commented:
@DylanCLi I see that you're still iterating on this PR a bit! Just let us know when you think it's ready for another review pass!

cbaker6 commented:
Hi @erik-apple I think @DylanCLi has this ready for review. I believe he finished his initial changes

erik-apple commented:
@DylanCLi Do you think you could add this into the Catalog app as well? That will make it easier to review and it will help others find it more easily once we get it merged!

DylanCLi commented:
Yeah, I just added it to the Catalog app.

erik-apple commented:
@DylanCLi I updated your PR to point into the new `main` branch. Main will replace master as the primary branch going forward!

gavirawson-apple commented on CareKitUI/CareKitUI/iOS/Task/Collection/OCKSlider.swift:
```diff
@@ -0,0 +1,235 @@
+
+import SwiftUI
+
+public struct OCKSlider: View {
```
```suggestion
public struct Slider: View {
```

gavirawson-apple commented on CareKitUI/CareKitUI/iOS/Task/Collection/OCKSliderStyle.swift:
```diff
@@ -0,0 +1,46 @@
+    case CareKitSlider(OCKSliderDimensions)
+    
+    /// An OCK Slider that uses the body of a system slider.
+    case UISlider
```
```suggestion
    case system
```

gavirawson-apple commented on CareKitUI/CareKitUI/iOS/Task/Collection/OCKSliderStyle.swift:
````diff
@@ -0,0 +1,46 @@
+    ///      +––––––––––––––––––––––––––––––––––+
+    ///      |      '      '      '      '      |
+    /// ```
+    case CareKitSlider(OCKSliderDimensions)
````
```suggestion
    case ticked
```

gavirawson-apple commented on CareKitUI/CareKitUI/iOS/Task/SliderTaskView.swift:
```diff
@@ -0,0 +1,243 @@
+    init(title: Text, detail: Text? = nil, instructions: Text? = nil, isComplete: Bool,
+         initialValue: CGFloat? = nil, value: Binding<CGFloat>, range: ClosedRange<CGFloat>, step: CGFloat = 1,
+         minimumImage: Image? = nil, maximumImage: Image? = nil, sliderStyle: OCKSliderStyle, isSliderTop: Bool = false,
+         action: @escaping (Double) -> Void) {
```
The type of the `value` accessible in the action should match the type of the `value` passed in during init.

gavirawson-apple commented on CareKit/CareKit/iOS/Task/SynchronizedSliderTaskView.swift:
```diff
@@ -0,0 +1,56 @@
+import SwiftUI
+
+@available(iOS 14.0, *)
+public struct SynchronizedSliderTaskView<Controller: OCKSliderTaskController, SliderTaskView: View>: View {
```
You can use the `SynchronizedTaskView ` instead of creating a new struct specifically for the slider task view.

gavirawson-apple commented on CareKitUI/CareKitUI/iOS/Task/Buttons/OCKSliderButton.swift:
```diff
@@ -0,0 +1,61 @@
+
+import SwiftUI
+
+public struct OCKSliderButton: View {
```
```suggestion
struct SliderButton: View {
```

gavirawson-apple commented on CareKitUI/CareKitUI/iOS/Task/SliderTaskView.swift:
````diff
@@ -0,0 +1,243 @@
+///     |                                                       |
+///     +-------------------------------------------------------+
+/// ```
+public struct SliderTaskView<Header: View, SliderView: View>: View {
````
This view saves a new outcome value each time the button is tapped. That behavior will be made clear to the user if we add the word `log` into the name of the view.

gavirawson-apple commented on CareKitUI/CareKitUI/iOS/Task/Collection/OCKSliderStyle.swift:
```diff
@@ -0,0 +1,46 @@
+}
+
+/// The dimensions that determine the size and frame of an OCKSlider with a filler style
+public struct OCKSliderDimensions {
```
We can use reasonable defaults instead of making this a configurable part of the API.

gavirawson-apple commented on CareKit/CareKit/iOS/Task/SliderTaskView.swift:
```diff
@@ -0,0 +1,177 @@
+public struct SliderTaskView<Header: View, SliderView: View>: View {
+    
+    private typealias TaskView = SynchronizedSliderTaskView<OCKSliderTaskController, CareKitUI.SliderTaskView<Header, SliderView>>
+
```
You can pop the `@State private var value: CGFloat` in this view.

gavirawson-apple commented on CareKitUI/CareKitUITests/Task/TestSliderTaskView.swift:
```diff
@@ -0,0 +1,39 @@
+
+class TestSliderTaskView: XCTestCase {
+    
+    @State var value: CGFloat = 0
```
`@State` should only be used in a SwiftUI view. You can test the slider using a mock slider view, passing in a constant for the `value`:

```swift
struct SliderView: View {

    @Binding
    var value: CGFloat
}

struct MockSliderView {

    let value: CGFloat

    var body: some View {
        SliderView(value: .constant(value))
    }
}
```

gavirawson-apple commented on CareKitUI/CareKitUI/iOS/Task/SliderTaskView.swift:
```diff
@@ -0,0 +1,243 @@
+    
+    /// Create an instance.
+    /// - Parameter instructions: Instructions text to display under the header.
+    /// - Parameter isComplete; True if the button under the slider is in the completed.
```
What effect does this parameter have on the view? This view allows you to log multiple outcome values for a single task, so does it matter if the task is completed, given you can always log additional outcomes?

DylanCLi replied:
Originally, I did not have logging in mind as its primary function. I removed `isComplete` from all of the Slider task files.

gavirawson-apple reviewed (changes requested):
Looking great Dylan! You're a rockstar!

gavirawson-apple commented:
@DylanCLi Thank you tons for all your hard work on this PR!

Let us know the next steps you would like to take with this PR. We're super excited about this contribution and want to get your work merged in! We'd be happy to work on the final refinements on our end if that makes your life easier.

DylanCLi commented:
@gavirawson-apple Thank you! I am still struggling with getting the `setEvent` changes to still save and show and also with moving the `@State var value` to the other file.

gavirawson-apple commented:
> I am still struggling with getting the setEvent changes to still save

Just to clarify, are you having trouble persisting the outcomes from the slider log to the store?

DylanCLi commented:
The issue came when trying to implement this https://github.com/carekit-apple/CareKit/pull/479#discussion_r482252145 by calling the existing setEvent function with `!values.isEmpty` in the `isComplete` parameter. 
```swift
open func setEvent(atIndexPath indexPath: IndexPath,
                   values: [OCKOutcomeValue], 
                   completion: ((Result<OCKAnyOutcome, Error>) -> Void)?) {
        setEvent(atIndexPath:isComplete: !values.isEmpty, completion)
}
```

cbaker6 commented on CareKit/CareKit/Shared/Task/Controller/OCKTaskController.swift:
```diff
@@ -299,6 +299,48 @@ open class OCKTaskController: ObservableObject {
+    ///   - isComplete: True if the event is complete.
+    ///   - values: Array of OCKOutcomeValue
+    ///   - completion: Result after setting the completion for the event.
+    open func setEvent(atIndexPath indexPath: IndexPath, isComplete: Bool, values: [OCKOutcomeValue], completion: ((Result<OCKAnyOutcome, Error>) -> Void)?) {
```
Fixed

cbaker6 commented on CareKit/CareKit/iOS/Task/Controllers/OCKSliderLogTaskController.swift:
```diff
@@ -0,0 +1,64 @@
+        guard !taskEvents.isEmpty else { return nil }
+
+        if let foundValue = taskEvents.first?.first?.outcome?.values.last?.numberValue?.doubleValue {
+            value = foundValue
```
This issue is still popping up. Basically, the value shown on the slider and button is sometimes outdated. It behaves like a threading issue where it will show the previous value when the button is pressed, but not the current.

This can be seen when testing the slider in Catalog. Sometimes all of the sliders will jump to the wrong value.

cbaker6 commented on CareKit/CareKit/iOS/Task/SynchronizedSliderLogTaskView.swift:
```diff
@@ -0,0 +1,56 @@
+                value = controller.value
+            }
+            .onReceive(controller.$value) { updatedValue in
+                value = updatedValue
```
This attempts to subscribe and receive the updated value from the viewModel.

DylanCLi commented:
@gavirawson-apple We were wondering since this is a slider log there might need to be changes to the button, which currently does not visually respond to being tapped. An idea is to have the button highlight for a a few seconds to show that it has logged a value, and then return to normal to allow for further logging. Is this something that would work? And how would we best implement this?

cbaker6 commented:
@gavirawson-apple originally @DylanCLi had the slider acting more like `OCKInstructionTask` by only saving one instruction and freezing the slider, allowing the user to delete the outcome by pressing the button again.

The current ways is more like `OCKButtonLog`. Is there a way to get the value log list view from `OCKButtonLog` into the SliderLogView? If so, it would look great

gavirawson-apple commented:
> We were wondering since this is a slider log there might need to be changes to the button, which currently does not visually respond to being tapped. An idea is to have the button highlight for a a few seconds to show that it has logged a value, and then return to normal to allow for further logging. Is this something that would work? And how would we best implement this?

@DylanCLi  Great question, the UI can get tricky when multiple values can be logged. Here's a mockup of an idea you can run with:

<img src="https://user-images.githubusercontent.com/51756298/100648597-00b2e600-32f6-11eb-976d-9d4eb4f51f2a.png">

There are a couple of nice UX aspects here:

- A value cannot be logged if the user hasn't touched the slider.

- Logging a value will reset the slider back to the original state. 

- The latest logged value is shown on the bottom of the card. Tapping the value can pop up a modal with all of the logged values. That way, the card won't grow vertically as more values are logged. (Don't worry about the modal view for your PR!)

I would recommend splitting up this PR to help us give you the best possible feedback and get it merged in! A nice idea might be to start with the CareKitUI view, and a single default style for the slider.

> originally @DylanCLi had the slider acting more like OCKInstructionTask by only saving one instruction and freezing the slider, allowing the user to delete the outcome by pressing the button again.

> The current ways is more like OCKButtonLog. Is there a way to get the value log list view from OCKButtonLog into the SliderLogView? If so, it would look great

@cbaker6 The views used in the `OCKButtonLog` are public, but I would recommend avoiding them and re-writing the views in SwiftUI. We would also like to avoid having the card expand vertically as more values are logged - check out the mockup above and let me know if that helps!

DylanCLi commented:
Hey @gavirawson-apple! I made some changes to match your mockup, as well as making a couple additions:
- I added an option to make a gradient color from an array of Colors similar to ResearchKit's slider (the system style slider in the top left has the gradient bleed through the part that is supposed to be grey, I couldn't find a way to fix this)
- I updated the ticked slider's appearance

<p float="left">
<img src="https://user-images.githubusercontent.com/56743995/101050803-c1350580-3552-11eb-934a-493b90f52ca0.gif" width="175" height="376">
<img src="https://user-images.githubusercontent.com/56743995/101050804-c2663280-3552-11eb-9eeb-24380a7fd89e.gif" width="175" height="376">
</p>

gavirawson-apple commented:
That looks great @DylanCLi! Thanks for all the hard work.

Can you submit a PR with just the CareKitUI implementation? That will help reduce the size of the PR so we can merge it in!

gavirawson-apple commented:
Thanks again for all the hard work @DylanCLi! Can we close this PR out in favor of #529?
